### ISSUE TITLE
DialErrorLimit create error on mismatching plugin versions

### ISSUE THREAD

**Issue by adrianriobo** [pulumi/pulumi-command#187]
### What happened?

Recently #180 introduced a new parameter `DialErrorLimit` that parameter has a default value which [prevents from panic](https://github.com/pulumi/pulumi-command/blob/e0d039d1c343e83b779f7a8fbf13726d6ef49ba6/provider/pkg/provider/remote/connection.go#L119) 

The main problem is that default value is set on the client side and the check is done on the plugin side. So in a situation when code is based on previous versions and the plugin is dynamically installed (based on latest version) the plugin will try to access a value which does not exist:

```bash
77    panic: runtime error: invalid memory address or nil pointer dereference
78    [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xdfa090]
79    goroutine 34 [running]:
80    [github.com/pulumi/pulumi-command/provider/pkg/provider/remote.(*Connection).Dial.func1(0x0](http://github.com/pulumi/pulumi-command/provider/pkg/provider/remote.%28*Connection%29.Dial.func1%280x0), 0xc00010f068?)
81        /home/runner/work/pulumi-command/pulumi-command/provider/pkg/provider/remote/connection.go:119 +0x1d0
```

### Expected Behavior

When a newer plugin version is installed it should not break the usage with old client code, so new parameters should check if they are not passed from the client and if that is the case the default value should be set at the plugin side.

### Steps to reproduce

Pick code for command provider version `0.7.0` and let pulumi handle the plugin installation (it will pick `0.7.1`) it will create the above error 

### Output of `pulumi about`

N/A

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by adrianriobo** [comment]
I guess it is always a best practice to ensure the matching of versions between provider client side and plugin. In any case I will create a patch for this trying to set the default value if we get a nil flrom the client.

**Comment by AaronFriel** [comment]
Hey @adrianriobo, I'm curious how this could arise? Pulumi will always select the plugin version that corresponds to the SDK version, e.g.:

```sh
pulumi init typescript
npm install --save @pulumi/command@0.7.1
# add command to index.ts
pulumi up
# will install command at 0.7.1
```

Is this `qenvs` tool managing plugins by putting them directly on the path? It's a workflow we use for development of plugins, but not one we recommend for general use.

**Comment by adrianriobo** [comment]
Hi @AaronFriel nop `qenvs` uses `automation api` and only install explicitly the `aws` plugin. 

What I notice is I have been using the command plugin on version `v0.7.0` for some time and `qenvs` is packed on a container `qenvs:v0.0.2`. As plugin is installed during runtime what I notice is right after version `v0.7.1` of the plugin was released the container my container start to failed...as you saw on the description.

So my only guess is qenvs client side of the command provider is based on `v0.7.0` and when the container is executed as the plugin was not installed during container build it downloads the latest version of the plugin so it breaks. 

:thinking:  although according to your explanation that should not be the case.

So if this is happening and it is not intended it a bug on pulumi side and I should not need to handle the syncs between client and plugin versions for providers WDYT?

**Comment by iwahbe** [comment]
Hi @adrianriobo. 

I agree with @AaronFriel that we should always match SDK and plugin version, but there are several situations where the engine can matche a plugin with the wrong version, including Automation API contexts. 

If you are still planning to supply a patch, we would be happy to take it. Otherwise, I can add it.

**Comment by adrianriobo** [comment]
@iwahbe yeah I will send it, I was waiting on a response ... but from your comment I assume on an Automation API context it is expected to handle somehow the match of the versions. 

So yeah on one side I will open a PR to ensure backwards version mismatch compatibility and I will create some helper classes on my project to ensure versioning within Automation API.
